src/main/java/com/hubspot/imap/utils/CRLFOutputStream.java

```diff
@@ -0,0 +1,59 @@
+package com.hubspot.imap.utils;
+
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+public class CRLFOutputStream extends FilterOutputStream {
+
+  private static final int CR = '\r';
+  private static final int LF = '\n';
+
+  private int previous = 0;
+
+  public CRLFOutputStream(OutputStream out) {
+    super(out);
+  }
+
+  public void write(int b) throws IOException {
+    if (b == CR) {
+      writeCRLF();
+    } else if (b == LF && previous != CR) {
+      writeCRLF();
+    } else {
+      out.write(b);
+    }
+
+    previous = b;
+  }
+
+  public void write(byte b[], int off, int len) throws IOException {
+    int begin = off;
+    int adjustedLen = len + off;
+
+    for (int i = off; i < adjustedLen; i++) {
+      int current = b[i];
+      if (current == CR) {
+        out.write(b, begin, i - begin);
+        writeCRLF();
+        begin = i + 1;
```

Comment:
`begin++`

Reply:
I don't think that would work - begin (offset) needs to be set to where the current index is, then plus one so that we start writing the bytes after the CRLF

Reply:
but you already set `begin` to the current off on line 31

Reply:
If we start off with `write(byte [...], 0, 20)`, we will iterate through the byte array until we see our first CR or LF (let's say at index 10). So then we write `out.write(b, begin, i - begin);` or `out.write([...], 0, 10 - 0);`. We then need `begin` to be 11 so that the next time we see a CR or LF (let's say at index 20), and call `out.write(b, begin, i - begin);`, we will write `out.write([...], 11, 20 - 11);` to write the remainder of the bytes. Otherwise, if I did `begin++` then we would end up with `out.write([...], 1, 20 - 1);` after finding the second CR or LF and in our outputstream wind up with the first 10 bytes followed by the 2nd through 10th bytes again and then the remaining bytes.